src/service/action-impl.js

```diff
@@ -982,32 +1023,46 @@ function argValueForTokens(tokens) {
 }
 
 /**
- * Dereferences expression args in `args` using data in `event`.
+ * Dereferences expression args in `args` using values in data.
  * @param {?ActionInfoArgsDef} args
- * @param {?ActionEventDef} event
+ * @param {!JsonObject|null|undefined} data
+ * @param {string=} opt_key The value to use in evaluating the argument expression.
+ *    This will be parent property of provided data and the root property
+ *    in the expression to evaluate against.
  * @return {?JsonObject}
  * @private
- * @visibleForTesting
  */
-export function dereferenceExprsInArgs(args, event) {
+export function dereferenceArgsVariables(args, data, opt_key) {
   if (!args) {
     return args;
   }
-  const data = dict();
-  if (event && getDetail(/** @type {!Event} */ (event))) {
-    data['event'] = getDetail(/** @type {!Event} */ (event));
+  let dataStore = dict({});
+
+  if (opt_key) {
+    if (data && data[opt_key]) {
+      dataStore = data;
+    } else {
+      dataStore[opt_key] = data;
+    }
+  } else if (data) {
+    dataStore = data;
   }
   const applied = map();
   Object.keys(args).forEach(key => {
     let value = args[key];
     if (typeof value == 'object' && value.expression) {
       const expr =
         /** @type {ActionInfoArgExpressionDef} */ (value).expression;
-      const exprValue = getValueForExpr(data, expr);
+      const exprValue = getValueForExpr(dataStore, expr);
       // If expr can't be found in data, use null instead of undefined.
       value = (exprValue === undefined) ? null : exprValue;
     }
-    applied[key] = value;
+    if (!!value
+        && Object.values(dataStore).length > 0 && !!dataStore[value]) {
+      applied[key] = dataStore[value];
```

Comment:
Please add a comment explaining why we need this case.

Reply:
I still think we could use a comment here. E.g.

```
// Only JSON expression strings that contain dereferences (e.g. `foo.bar`)
// are processed as ActionInfoArgExpressionDef. We also support dereferencing
// strings like `foo` iff there is a corresponding key in `data`. Otherwise, `foo`
// is treated as a string "foo".
```